libminifi/src/ConfigurationListener.cpp

```diff
@@ -0,0 +1,130 @@
+#include "FlowController.h"
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+#include <string>
+#include <memory>
+#include <utility>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+
+void ConfigurationListener::start() {
+  if (running_)
+    return;
+
+  pull_interval_ = 60 * 1000;
+  std::string value;
+  // grab the value for configuration
+  if (configure_->get(Configure::nifi_configuration_listener_pull_interval,
+      value)) {
+    core::TimeUnit unit;
+    if (core::Property::StringToTime(value, pull_interval_, unit)
+        && core::Property::ConvertTimeUnitToMS(pull_interval_, unit,
+            pull_interval_)) {
+      logger_->log_info("Configuration Listener pull interval: [%d] ms",
+           pull_interval_);
+    }
+  }
+
+  std::string clientAuthStr;
+  if (configure_->get(Configure::nifi_configuration_listener_need_ClientAuth, clientAuthStr)) {
+    org::apache::nifi::minifi::utils::StringUtils::StringToBool(clientAuthStr, this->need_client_certificate_);
+  }
+
+  if (configure_->get(
+          Configure::nifi_configuration_listener_client_ca_certificate,
+      this->ca_certificate_)) {
+    logger_->log_info("Configuration Listener CA certificates: [%s]",
+        this->ca_certificate_.c_str());
+  }
+
+  if (this->need_client_certificate_) {
+    std::string passphrase_file;
+
+    if (!(configure_->get(
+        Configure::nifi_configuration_listener_client_certificate, this->certificate_)
+        && configure_->get(Configure::nifi_configuration_listener_private_key,
+            this->private_key_))) {
+      logger_->log_error(
+          "Certificate and Private Key PEM file not configured for configuration listener, error: %s.",
+          std::strerror(errno));
+    }
+
+    if (configure_->get(
+        Configure::nifi_configuration_listener_client_pass_phrase,
+        passphrase_file)) {
+      // load the passphase from file
+      std::ifstream file(passphrase_file.c_str(), std::ifstream::in);
+      if (file.good()) {
+        this->passphrase_.assign((std::istreambuf_iterator<char>(file)),
+            std::istreambuf_iterator<char>());
+        file.close();
+      }
+    }
+
+    logger_->log_info("Configuration Listener certificate: [%s], private key: [%s], passphrase file: [%s]",
+            this->certificate_.c_str(), this->private_key_.c_str(), passphrase_file.c_str());
+  }
+
+  thread_ = std::thread(&ConfigurationListener::threadExecutor, this);
+  thread_.detach();
+  running_ = true;
+  logger_->log_info("%s ConfigurationListener Thread Start", type_.c_str());
+}
+
+void ConfigurationListener::stop() {
+  if (!running_)
+    return;
+  running_ = false;
+  if (thread_.joinable())
+    thread_.join();
+  logger_->log_info("%s ConfigurationListener Thread Stop", type_.c_str());
+}
+
+void ConfigurationListener::run() {
+  int64_t interval = 0;
+  while (running_) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
is the 100ms sleep to allow this object to check for interruptibility? If that is the case could you use a condition variable with a wait_for and avoid the checks, sleep, and regular wakeup?